libvisual/libvisual/lv_video.cpp

```diff
@@ -292,6 +293,110 @@ namespace LV {
       return true;
   }
 
+  #ifdef LV_ENABLE_TESTS
+  bool Video::validate() const
+  {
+      // Validate dimensions.
+
+      if (m_impl->width <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero or negative width (%d).", m_impl->width);
+          return false;
+      }
+
+      if (m_impl->height <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero of negative height (%d).", m_impl->height);
+          return false;
+      }
+
+      if (m_impl->pitch <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero of negative pitch (%d).", m_impl->pitch);
+          return false;
+      }
+
+      // Validate extents.
+
+      if (m_impl->width != m_impl->extents.width) {
+          visual_log (VISUAL_LOG_ERROR, "Width (%d) and width of extents (%d) do not agree.",
+                      m_impl->width, m_impl->extents.width);
+          return false;
+      }
+
+      if (m_impl->height != m_impl->extents.height) {
+          visual_log (VISUAL_LOG_ERROR, "Height (%d) and height of extents (%d) do not agree.",
+                      m_impl->height, m_impl->extents.height);
+          return false;
+      }
+
+      // Validate parent-child relations.
+
+      if (m_impl->parent) {
+          auto parent_impl = m_impl->parent->m_impl.get ();
+          if (!parent_impl->extents.contains (m_impl->extents)) {
+              visual_log (VISUAL_LOG_ERROR, "Sub-video is not fully contained by parent.");
+              return false;
+          }
+      }
+
+      if (m_impl->depth != VISUAL_VIDEO_DEPTH_GL) {
+          // Validate depth and bytes per pixel consisteency.
```

Comment:
Typo "consisteency" here^^

Reply:
@hartwork

> @kaixiong a few comments on details below. Regarding the big picture, I'm wondering if adding the testing code and using runtime logging is a bug or a feature. Is the key driver to be able to access internals easily or is there something else? Normally I would have expected to have the test code go to dedicated files and to use test-specific "assertions" (e.g. thinking gtest) rather than libvisual logging. What do you think?

In general, I agree.

`LV::Video::has_same_content()` is potentially useful by apps, so I made it available.

`LV::Video::validate()` is purely for internal use and testing. The function needs access to internals, and so cannot be moved to the testing code. It also comes in handy within Core itself during debugging.

We could make `validate()` part of the API but that seems wrong because `validate()` checks class invariants that will never be violated with the eventual API (where mutability of attributes is removed).

The logging is a workaround at the moment because to properly return the failed test result(s) from the function, we would need a sum/variant type (think Rust-style error handling), something I don't want to introduce just yet.

As I'm typing this comment, I realise there's a flaw in the build configuration. The examples and tools are also able to access `validate()`, so `LV_ENABLE_TESTS` has to be defined in only tests/ and libvisual/.